subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java

```diff
@@ -506,6 +637,67 @@ public ProviderInternal<T> toProvider() {
         public ExecutionTimeValue<T> withChangingContent() {
             return new FixedExecutionTimeValue<>(value, true);
         }
+
+        @Override
+        public ExecutionTimeValue<T> withSideEffect(SideEffect<? super T> sideEffect) {
+            return new FixedWithSideEffectExecutionTimeValue<>(value, changingContent, sideEffect);
+        }
+    }
+
+    class FixedWithSideEffectExecutionTimeValue<T> extends FixedExecutionTimeValue<T> {
+
+        private final SideEffect<? super T> sideEffect;
+
+        private FixedWithSideEffectExecutionTimeValue(T value, boolean changingContent, SideEffect<? super T> sideEffect) {
+            super(value, changingContent);
+            this.sideEffect = sideEffect;
+        }
+
+        @Override
+        public SideEffect<? super T> getSideEffect() {
+            return sideEffect;
+        }
+
+        @Override
+        public ExecutionTimeValue<T> withSideEffect(SideEffect<? super T> sideEffect) {
+            // avoid capturing the value wrapper
+            SideEffect<? super T> prevSideEffect = this.sideEffect;
+            SideEffect<? super T> chainedSideEffect = it -> {
+                prevSideEffect.accept(it);
+                sideEffect.accept(it);
+            };
+            return new FixedWithSideEffectExecutionTimeValue<>(super.getFixedValue(), hasChangingContent(), chainedSideEffect);
+        }
+
+        @Override
+        public ExecutionTimeValue<T> withChangingContent() {
+            return new FixedWithSideEffectExecutionTimeValue<>(super.getFixedValue(), true, sideEffect);
+        }
+
+        @Override
+        public T getFixedValue() {
+            runSideEffect();
```

Comment:
Looking at the usages of this method, it doesn't seem to make sense to run the side effect here. The usages show that collection properties would not emit the side effect when queried.

Reply:
You are maybe right that we shouldn't run the side effect here.

Looking at the usages of `getFixedValue` made me realize there are more places in the code that potentially need to preserve the side effects. Like this one:
https://github.com/gradle/gradle/blob/aab2df65b07e178396c9dc39a2d1e088d7522080/subprojects/model-core/src/main/java/org/gradle/api/internal/provider/Collectors.java#L118-L120

Reply:
I migrated most of the places using the `getFixedValue` to preserve side effects. I believe there is only one more left, that is a tricky one:
https://github.com/gradle/gradle/pull/21373/files#diff-acd0c9c9cb69fc3c944be33befdde036f5f448a58517dd1b140fb25ac9a0a225R345